### PR TITLE
Move 'close before delete' preconditions downstream

### PR DESCRIPTION
**Pull request by MichaelSnowden**
This diff does the refactoring that https://github.com/temporalio/temporal/issues/3368 proposes.

### CODE REVIEW

File: proto/internal/temporal/server/api/persistence/v1/executions.proto [hunk 1]
```diff
@@ -172,6 +172,7 @@ message TransferTaskInfo {
     // (-- api-linter: core::0140::prepositions=disabled
     //     aip.dev/not-precedent: "after" is used to indicate sequence of actions. --)
     bool delete_after_close = 15;
+    int64 close_transfer_task_id = 16;
```
Comment (yux0) [inline proto/internal/temporal/server/api/persistence/v1/executions.proto:175]: You don't need this one in transfer deletion task. This data is accessible in the mutable state in transfer task

File: service/history/visibilityQueueTaskExecutor.go [hunk 2]
```diff
@@ -439,6 +452,37 @@ func (t *visibilityQueueTaskExecutor) processDeleteExecution(
...
 
+func (t *visibilityQueueTaskExecutor) isCloseVisibilityTaskDone(task *tasks.DeleteExecutionVisibilityTask) (bool, error) {
+	namespaceEntry, err := t.shard.GetNamespaceRegistry().GetNamespaceByID(namespace.ID(task.NamespaceID))
+	if err != nil {
+		return false, err
+	}
+	currentClusterName := t.shard.GetClusterMetadata().GetCurrentClusterName()
+	nsActive := namespaceEntry.ActiveInCluster(currentClusterName)
+	closeVisibilityTaskID := task.CloseVisibilityTaskID
+	// taskID == 0 if workflow still running in passive cluster or closed before this field was added (v1.17).
```
[the remark this answers is not in the document]
Comment (MichaelSnowden) [reply service/history/visibilityQueueTaskExecutor.go]: Done

File: service/history/visibilityQueueTaskExecutor.go [hunk 3]
```diff
@@ -439,6 +452,37 @@ func (t *visibilityQueueTaskExecutor) processDeleteExecution(
...
+	// check if close execution visibility task is completed
+	if !nsActive {
+		return true, nil
+	}
+	if closeVisibilityTaskID == 0 {
+		return true, nil
+	}
```
[the remark this answers is not in the document]
Comment (MichaelSnowden) [reply service/history/visibilityQueueTaskExecutor.go]: Done

File: service/history/workflow/delete_manager.go [hunk 4]
```diff
@@ -302,6 +253,32 @@ func (m *DeleteManagerImpl) deleteWorkflowExecutionInternal(
 	return nil
 }
 
+func (m *DeleteManagerImpl) isTaskAcked(namespaceID namespace.ID, workflowKey definition.WorkflowKey, taskId int64) (bool, error) {
```
[the remark this answers is not in the document]
Comment (MichaelSnowden) [reply service/history/workflow/delete_manager.go]: Done

File: proto/internal/temporal/server/api/persistence/v1/executions.proto (hunk 1 above)
Comment (MichaelSnowden) [reply proto/internal/temporal/server/api/persistence/v1/executions.proto]: Done

**Pull request closed without merging**